control-base/src/main/java/fi/nls/oskari/control/feature/AbstractFeatureHandler.java

```diff
@@ -112,91 +122,131 @@ protected boolean isAllowedGeomType(String type) {
+            for (int i = 0; i < jsonArray.length(); i++) {
+                JSONObject property = jsonArray.getJSONObject(i);
+                if(property.getString("value").isEmpty()) {
+                    continue;
+                }
+                feature.addProperty(property.getString("key"), property.getString("value"));
+            }
+        }
+
+        if (jsonObject.has("geometries")) {
+            JSONObject geometries = jsonObject.getJSONObject("geometries");
+            String geometryType = geometries.getString("type");
+            if(!isAllowedGeomType(geometryType)) {
+                throw new ActionParamsException("Invalid geometry type: " + geometryType);
+            }
+            JSONArray data = geometries.getJSONArray("data");
+            feature.setGeometry(getGeometry(geometryType, data, getSrid(srsName, 3067)));
+        }
+        return feature;
+    }
+
+    protected int getSrid(String srsName, int defaultValue) {
+        if (srsName != null) {
+            int i = srsName.lastIndexOf(':');
+            if (i > 0) {
+                srsName = srsName.substring(i + 1);
+            }
+            try {
+                return Integer.parseInt(srsName);
+            } catch (NumberFormatException ignroe) {}
+        }
+        return defaultValue;
+    }
+
+
+    protected Geometry getGeometry(String geometryType, JSONArray data, int srid) throws ActionParamsException, JSONException {
         if ("multipoint".equals(geometryType)) {
-            return getMultipoint(srsName, data);
+            return getMultipoint(srid, data);
         } else if ("multilinestring".equals(geometryType)) {
-            return getMultiLineStringGeometries(srsName, data);
+            return getMultiLineStringGeometries(srid, data);
         } else if ("multipolygon".equals(geometryType)) {
-            return getMultiPolygonGeometries(srsName, data);
+            return getMultiPolygonGeometries(srid, data);
         }
         throw new ActionParamsException("Unknown type");
     }
 
-    protected String getMultipoint(String srsName, JSONArray data) throws JSONException {
-        StringWriter writer = new StringWriter();
-        writer.append("<gml:MultiPoint xmlns:gml=\"http://www.opengis.net/gml\" srsName=\"");
-        writer.append(srsName);
-        writer.append("\">");
+    protected Geometry getMultipoint(int srid, JSONArray data) throws JSONException {
+        List<Point> points = new ArrayList<>();
         for (int i = 0; i < data.length(); i++) {
-            writer.append("<gml:pointMember><gml:Point><gml:coordinates decimal=\".\" cs=\",\" ts=\" \">"
-                    + data.getJSONObject(i).getString("x") + "," + data.getJSONObject(i).getString("y")
-                    + "</gml:coordinates></gml:Point></gml:pointMember>");
+            Point point = gf.createPoint(new Coordinate(
+                    Double.parseDouble(data.getJSONObject(i).getString("y")),
```

Comment:
Axis order gets swapped here. This was probably only tested with a CRS that has axis order northing/easting (y/x, lat/lon)? GML3Writer writes the JTS Geometry as x/y, but here we read y to x and x to y swapping the order. Probably should just read x to x and y to y here and then swap the order if necessary (depending on `CoordinateReferenceSystem` we decode with CRS.decode) using `CoordinateSequenceFilter` (see https://stackoverflow.com/a/29634389)